app/controllers/concerns/application_shared.rb

```diff
@@ -15,4 +15,35 @@ def set_timezone
     # Reset timezone for the next thread
     Time.zone = default_timezone
   end
-end
\ No newline at end of file
+
+  private
+
+  # Get the allowed parameters from the Apipie documentation in the API controllers.
+  def safe_params
+    apipie_params = Apipie.app.get_method_description(controller_name, action_name).to_json[:params]
```

Comment:
> 1) Yes it'd be kind of what we have with action_permission, not sure about the exact percentages :). I'm not sure which controllers will need manual specification of the attributes, are you referring to controllers like UnattendedController which doesn't have an API equivalent?

Yeah, indeed.  It might be a more common problem in plugins where many have UI controllers without any corresponding API.

> For 2) I think we should probably trigger Apipie.app.reload_documentation, if that method is returning nil for something that clearly exists, it just means the docs have not been loaded.

Sounds expensive!